test/functional/p2p_segwit.py

```diff
@@ -1418,16 +1418,24 @@ def test_segwit_versions(self):
         temp_utxo.pop()  # last entry in temp_utxo was the output we just spent
         temp_utxo.append(UTXO(tx2.sha256, 0, tx2.vout[0].nValue))
 
-        # Spend everything in temp_utxo back to an OP_TRUE output.
+        # Spend everything in temp_utxo into an segwit v1 output.
         tx3 = CTransaction()
         total_value = 0
         for i in temp_utxo:
             tx3.vin.append(CTxIn(COutPoint(i.sha256, i.n), b""))
             tx3.wit.vtxinwit.append(CTxInWitness())
             total_value += i.nValue
         tx3.wit.vtxinwit[-1].scriptWitness.stack = [witness_program]
-        tx3.vout.append(CTxOut(total_value - 1000, CScript([OP_TRUE])))
+        tx3.vout.append(CTxOut(total_value - 1000, script_pubkey))
         tx3.rehash()
+
+        # First we test this transaction against fRequireStandard=true node
+        # making sure the txid is added to the reject filter
+        self.std_node.announce_tx_and_wait_for_getdata(tx3)
+        test_transaction_acceptance(self.nodes[1], self.std_node, tx3, with_witness=True, accepted=False, reason="bad-txns-nonstandard-inputs")
+        # Now the node will no longer ask for getdata of this transaction when advertised by same txid
+        self.std_node.announce_tx_and_wait_for_getdata(tx3, timeout=5, success=False)
```

Comment:
I don't like that this is testing transaction acceptance by relying on our transaction download logic. It means that changes to transaction download (eg #19184) may cause unrelated tests to fail, or (perhaps worse) make it such that tests are no longer actually testing what we want them to test.

Unfortunately, we have no way to directly query a node's recent rejects filter, so we do need to test this indirectly somehow. Here's an alternative that uses the debug log to determine why the transaction has been rejected. It's not ideal, but I think it's better than testing via a separate subcomponent (tx download):

```diff
diff --git a/test/functional/p2p_segwit.py b/test/functional/p2p_segwit.py
index 519fb0438f..e0c671f55d 100755
--- a/test/functional/p2p_segwit.py
+++ b/test/functional/p2p_segwit.py
@@ -123,13 +123,14 @@ def get_virtual_size(witness_block):
     vsize = int((3 * base_size + total_size + 3) / 4)
     return vsize
 
-def test_transaction_acceptance(node, p2p, tx, with_witness, accepted, reason=None):
+def test_transaction_acceptance(node, p2p, tx, with_witness, accepted, reason=None, non_reason=None):
     """Send a transaction to the node and check that it's accepted to the mempool
 
     - Submit the transaction over the p2p interface
     - use the getrawmempool rpc to check for acceptance."""
     reason = [reason] if reason else []
-    with node.assert_debug_log(expected_msgs=reason):
+    non_reason = [non_reason] if non_reason else []
+    with node.assert_debug_log(expected_msgs=reason, unexpected_msgs=non_reason):
         p2p.send_and_ping(msg_tx(tx) if with_witness else msg_no_witness_tx(tx))
         assert_equal(tx.hash in node.getrawmempool(), accepted)
 
@@ -1418,7 +1419,7 @@ class SegWitTest(BitcoinTestFramework):
         temp_utxo.pop()  # last entry in temp_utxo was the output we just spent
         temp_utxo.append(UTXO(tx2.sha256, 0, tx2.vout[0].nValue))
 
-        # Spend everything in temp_utxo into an segwit v1 output.
+        # Spend everything in temp_utxo into a segwit v1 output.
         tx3 = CTransaction()
         total_value = 0
         for i in temp_utxo:
@@ -1433,8 +1434,9 @@ class SegWitTest(BitcoinTestFramework):
         # making sure the txid is added to the reject filter
         self.std_node.announce_tx_and_wait_for_getdata(tx3)
         test_transaction_acceptance(self.nodes[1], self.std_node, tx3, with_witness=True, accepted=False, reason="bad-txns-nonstandard-inputs")
-        # Now the node will no longer ask for getdata of this transaction when advertised by same txid
-        self.std_node.announce_tx_and_wait_for_getdata(tx3, timeout=5, success=False)
+        # If we send the transaction again, it'll be rejected by the recent
+        # rejects filter before we try to accept it to mempool
+        test_transaction_acceptance(self.nodes[1], self.std_node, tx3, with_witness=True, accepted=False, non_reason="bad-txns-nonstandard-inputs")
 
         # Spending a higher version witness output is not allowed by policy,
         # even with fRequireStandard=false.
```

Reply:
> Unfortunately, we have no way to directly query a node's recent rejects filter

Would it be better to have a unit test that calls ATMP and then checks that the txid made it into the reject filter?

Reply:
I'm not sure if that counts as a unit test? ATMP and the reject filter are in different components.

Reply:
Not sure it's "better", I still think the best thing would be to add log lines in case of rejection due to `recentRejects` and scan for that line in logs.

Reply:
> I still think the best thing would be to add log lines in case of rejection due to recentRejects and scan for that line in logs.

This seems reasonable, and very easy to implement.

Reply:
After some noodling I actually think a simple timeout `average_wait_for_typical_getdata<<n<<GETDATA_TX_INTERVAL` is best, or a "unit" test.

I think 5 seconds is ok for now, someone can replace later with something better.

Reply:
> I actually think a simple timeout

Can you explain why you think that? (I'm not disagreeing, just trying to understand why you think relying on tx download is a better test for this tx acceptance change).

There seems to be some confusion around the concept of unit tests here. This is higher-level functionality spread across two different components (validation and net_processing). The only testing framework that we have that encompasses those two components is the functional test framework. As far as I'm aware, we have cpp tests for net_processing (eg denialofservice_tests.cpp) and validation (eg validation_block_tests.cpp), but not tests that span both.

Generally, the concept of unit tests is to test units of code in isolation (eg here we could test that validation returns `TX_INPUTS_NOT_STANDARD` when the inputs are non-standard). Functional test test higher-level system functionality (eg we don't redownload or revalidate transactions which have non-standard inputs).

Reply:
I was mostly annoyed that the various failure cases have to be handled on a case by case basis. One case is `AlreadyHave`, the other is not but the `getdata` is not sent because another peer has already had it requested. I'd rather write something that covers both properly.